File: drills/python/solutions/nc31_search_2d_matrix.py
def searchMatrix(matrix, target):
    # The rows are sorted and do not overlap, so the row a value could live in
    # is itself found by halving: compare the target against a row's ends.
    low, high = 0, len(matrix) - 1
    while low <= high:
        mid = (low + high) // 2
        if matrix[mid][-1] < target:
            low = mid + 1
        elif matrix[mid][0] > target:
            high = mid - 1
        else:
            return contains(matrix[mid], target)
    return False


def contains(row, target):
    low, high = 0, len(row) - 1
    while low <= high:
        mid = (low + high) // 2
        if row[mid] == target:
            return True
        if row[mid] < target:
            low = mid + 1
        else:
            high = mid - 1
    return False

File: drills/elixir/solutions/nc31_search_2d_matrix.exs
defmodule Solution do
  def search_matrix(matrix, target) do
    case find_row(List.to_tuple(matrix), target, 0, length(matrix) - 1) do
      nil -> false
      row -> contains?(List.to_tuple(row), target, 0, length(row) - 1)
    end
  end

  # The rows are sorted and do not overlap, so the row a value could live in is
  # itself found by halving: compare the target against a row's ends.
  defp find_row(_rows, _target, low, high) when low > high, do: nil

  defp find_row(rows, target, low, high) do
    mid = div(low + high, 2)
    row = elem(rows, mid)

    cond do
      List.last(row) < target -> find_row(rows, target, mid + 1, high)
      hd(row) > target -> find_row(rows, target, low, mid - 1)
      true -> row
    end
  end

  defp contains?(_row, _target, low, high) when low > high, do: false

  defp contains?(row, target, low, high) do
    mid = div(low + high, 2)

    cond do
      elem(row, mid) == target -> true
      elem(row, mid) < target -> contains?(row, target, mid + 1, high)
      true -> contains?(row, target, low, mid - 1)
    end
  end
end
